src/rpc/rawtransaction.cpp

```diff
@@ -218,6 +218,252 @@ UniValue getrawtransaction(const UniValue &params, bool fHelp)
+            "      \"hex\" : \"data\",       (string) The serialized, hex-encoded data for 'txid'\n"
+            "      \"txid\" : \"id\",        (string) The transaction id (same as provided)\n"
+            "      \"size\" : n,             (numeric) The transaction size\n"
+            "      \"version\" : n,          (numeric) The version\n"
+            "      \"locktime\" : ttt,       (numeric) The lock time\n"
+            "      \"vin\" : [               (array of json objects)\n"
+            "         {\n"
+            "           \"txid\": \"id\",    (string) The transaction id\n"
+            "           \"vout\": n,         (numeric) \n"
+            "           \"scriptSig\": {     (json object) The script\n"
+            "             \"asm\": \"asm\",  (string) asm\n"
+            "             \"hex\": \"hex\"   (string) hex\n"
+            "           },\n"
+            "           \"sequence\": n      (numeric) The script sequence number\n"
+            "         }\n"
+            "         ,...\n"
+            "        ],\n"
+            "      \"vout\" : [              (array of json objects)\n"
+            "         {\n"
+            "           \"value\" : x.xxx,            (numeric) The value in " +
+            CURRENCY_UNIT +
+            "\n"
+            "           \"n\" : n,                    (numeric) index\n"
+            "           \"scriptPubKey\" : {          (json object)\n"
+            "             \"asm\" : \"asm\",          (string) the asm\n"
+            "             \"hex\" : \"hex\",          (string) the hex\n"
+            "             \"reqSigs\" : n,            (numeric) The required sigs\n"
+            "             \"type\" : \"pubkeyhash\",  (string) The type, eg 'pubkeyhash'\n"
+            "             \"addresses\" : [           (json array of string)\n"
+            "               \"bitcoinaddress\"        (string) bitcoin address\n"
+            "               ,...\n"
+            "             ]\n"
+            "           }\n"
+            "         }\n"
+            "         ,...\n"
+            "        ],\n"
+            "      \"blockhash\" : \"hash\",   (string) the block hash\n"
+            "      \"confirmations\" : n,      (numeric) The confirmations\n"
+            "      \"time\" : ttt,             (numeric) The transaction time in seconds since epoch (Jan 1 1970 GMT)\n"
+            "      \"blocktime\" : ttt         (numeric) The block time in seconds since epoch (Jan 1 1970 GMT)\n"
+            "    },\n"
+            "    ...\n"
+            "  },\n"
+            "  ...\n"
+            "}\n"
+            "\nExamples:\n" +
+            HelpExampleCli("getrawtransactionssince", "\"hashblock\"") +
+            HelpExampleCli("getrawtransactionssince", "\"hashblock\" 1") +
+            HelpExampleCli("getrawtransactionssince", "\"hashblock\" 1 10") +
+            HelpExampleRpc("getrawtransactionssince", "\"hashblock\", 1 10"));
+
+    LOCK(cs_main);
```

Comment:
you are right that cs_main needs to be held here in case of a reorg, so that we return a set of blocks that is consistent.  However, note that a reorg could happen moments after this call -- the caller still needs to handle that case.

So one way to minimize cs_main time is only lock it while copying block hashes from chainActive to a temporary array.

Then you can unlock during the lengthy period where you read each block off of the disk and process it.

optionally, at the very end check that chainActive[lastblock] == last block in your array, and if it does not throw out your results (or part of them) and recalculate.